### ISSUE TITLE
Automatically increment the date part

### ISSUE THREAD

**Issue by href** [lgiordani/punch#13]
I'm using punch to build a package which includes a month part in its version number. Basically we do *major*.*month*.*build*.

For this I created the following config:

```python
__config_version__ = 1

GLOBALS = {
    'serializer': '{{year}}.{{month}}.{{build}}',
}

FILES = ['setup.py']

VERSION = [
    {
        'name': 'year',
        'type': 'date',
        'fmt': 'YY'
    },
    {
        'name': 'month',
        'type': 'date',
        'fmt': 'MM'
    },
    {
        'name': 'build',
        'type': 'integer',
        'start_value': 0
    }
]

VCS = {
    'name': 'git',
    'commit_message': "Release {{ new_version }}",
}
```

However, I noticed that the month part of the build will only update if I explicitly say so. That is, to create a new build at the beginning of the month I need to do the following:

`punch --part month`

While during the month I need to run:

`punch --part build`

In other words, I need to manually roll the month version over. Is there an option to always do this automatically? I would expect a date version part to automatically change when the date changes.

**Comment by lgiordani** [comment]
This is an interesting use case, thanks. The elements in the `VERSION` list are hierarchical so when you update `month` it automatically updates everything after it (`build` in this case).

You can swap `month` and `build` in the `VERSION` list, thus, to have the `month` part automatically managed by `--part build`, while in the `serializer` you can keep them in the current order.

This is the test (I'm going to add it to the suite asap)

``` python
import subprocess
import six
import pytest

if six.PY2:
    import mock
else:
    from unittest import mock

pytestmark = pytest.mark.slow

version_file_content = """
year = '2016'
month = '4'
build = '1'
"""

config_file_content = """
__config_version__ = 1

GLOBALS = {
    'serializer': '{{year}}.{{month}}.{{build}}',
}

FILES = ['README.md']

VERSION = [
    {
        'name': 'year',
        'type': 'date',
        'fmt': '%Y'
    },
    {
        'name': 'build',
        'type': 'integer',
        'start_value': 0
    },
    {
        'name': 'month',
        'type': 'date',
        'fmt': '%m'
    }
]
"""


def test_update_month(test_environment):
    test_environment.ensure_file_is_present("README.md", "Version 2016.4.1")

    test_environment.ensure_file_is_present("punch_version.py",
                                            version_file_content)

    test_environment.ensure_file_is_present("punch_config.py",
                                            config_file_content)

    system_year = subprocess.check_output(['date', '+%Y'])
    system_year = system_year.decode('utf8').replace('\n', '')

    system_month = subprocess.check_output(['date', '+%m'])
    system_month = system_month.decode('utf8').replace('\n', '')

    test_environment.call(["punch", "--part", "month"])

    assert test_environment.get_file_content("README.md") == \
        "Version 2016.{}.1".format(system_month)


def test_update_build(test_environment):
    test_environment.ensure_file_is_present("README.md", "Version 2016.4.1")

    test_environment.ensure_file_is_present("punch_version.py",
                                            version_file_content)

    test_environment.ensure_file_is_present("punch_config.py",
                                            config_file_content)

    system_month = subprocess.check_output(['date', '+%m'])
    system_month = system_month.decode('utf8').replace('\n', '')

    test_environment.call(["punch", "--part", "build"])

    assert test_environment.get_file_content("README.md") == \
        "Version 2016.{}.2".format(system_month)
```

**Comment by href** [comment]
Thanks for getting back to me so quickly. This almost gets me there, but I'm afraid I wasn't too clear in my description. When I say build, then I mean build per month.

So our versions in April/May might look like this:

- 2017.4.0
- 2017.4.1
- 2017.4.2
- 2017.5.0
- 2017.5.1

Is that something that can be solved as well? 😁

**Comment by lgiordani** [comment]
I see, sorry if I misunderstood the problem. With your initial setup what happens if you continuously issue `--part month`? Doesn't it reset build when the month changes? (Sorry, I cannot setup a test at the moment).

I expect the `--part month` to just produce the output you showed me, as `month` is higher in the hierarchy than `build`, just like the standard `minor` in semver.

**Comment by href** [comment]
You didn't misunderstand it actually, I described it poorly ;)

I tried running `--part month` multiple times with my original config (VERSION=year,month,build), which only works once:

- 2017.5.0
- 2017.5.0 with the following error

```python
Traceback (most recent call last):
  File "/Users/denis/.virtualenvs/onegov/bin/punch", line 215, in <module>
    uc.finish_release()
  File "/Users/denis/.virtualenvs/onegov/lib/python3.5/site-packages/punch/vcs_repositories/git_repo.py", line 79, in finish_release
    self._run([self.command, "branch", "-d", branch])
  File "/Users/denis/.virtualenvs/onegov/lib/python3.5/site-packages/punch/vcs_repositories/vcs_repo.py", line 53, in _run
    raise RepositorySystemError(error_message)
punch.vcs_repositories.exceptions.RepositorySystemError: An error occurred executing 'git branch -d heads/2017.5.0': error: branch 'heads/2017.5.0' not found.
```

With the adjusted config of VERSION=year,build,month, running `punch --part month` I get the same error. Presumably because the month doesn't change between invocations.

Ideally what I would want is for all both month and year to automatically take the current date into account. At the same time I would like to reset the build to 0 whenever the month changes.

I know this is a bit of an exotic example. If this is not something that punch should address then I'd actually be fine with it. I can also just write a little script that takes care of this for me.

**Comment by lgiordani** [comment]
Thanks for the explanation. I will work on it, I am pretty sure that the use case is not that specific, so the date version part can definitely behave like you describe.

**Comment by lgiordani** [comment]
Hey, take a look at #14 which contains a solution to your issue. The solution is general enough to allow me to cover other cases in the future, I think. You can use the branch in pip with this syntax

``` sh
pip install git+https://github.com/lgiordani/punch.git@feature/actions
```

I'm going to merge the PR in the next days.

**Comment by href** [comment]
Thanks, this is really great! I tried this on my end, but it didn't quite work as expected.

This is my current config:

```python
__config_version__ = 1

GLOBALS = {
    'serializer': '{{year}}.{{month}}.{{build}}',
}

FILES = ['setup.py']

VERSION = [
    {
        'name': 'year',
        'type': 'date',
        'fmt': 'YYYY'
    },
    {
        'name': 'month',
        'type': 'date',
        'fmt': 'MM'
    },
    {
        'name': 'build',
        'type': 'integer',
        'start_value': 0
    }
]

VCS = {
    'name': 'git',
    'commit_message': "Release {{ new_version }}",
}

ACTIONS = {
    'build': {
        'type': 'refresh',
        'refresh_fields': ['year', 'month'],
        'fallback_field': 'build'
    }
}
```

When I trigger it with my current version being set to '2017.4.28', then I end up with '2017.6.28'. What I would like to see is '2017.6.0'. As you can see the build number hasn't changed.

I wonder if it is because the month skipped from 4 to 6 (I didn't increment the month part of the version in may). In any case, if I understand your docs correctly I should see a reset of the build number to 0:

> So for the above configuration if the current version is `2017.01.4` on 31 January 2017 the command `punch --action mbuild` creates version `2017.01.5`, while on the 01 February 2017 it will create version `2017.02.0`.

This is my output of `punch --action build`:

`{'month': 4, 'year': 2017, 'build': 28} {'month': '6', 'year': '2017', 'build': 28}`

(Curiously, the year changed from an int to a string, or at least it's output that way)

**Comment by lgiordani** [comment]
I see, there are still issues! Probably my tests are wrong. Sorry =) I'll work on it, thanks for testing it so quickly.

**Comment by lgiordani** [comment]
It looks like my approach with actions is fine, but the action I implemented didn't really solve the problem :stuck_out_tongue: The current version should, I checked the new tests and they are correct now (at least I hope). May you please check if this version works?

The action changed, so check the documentation. I'm not 100% satisfied by the name of the action and the parameters, but for the moment I think I can live with that :see_no_evil:

**Comment by href** [comment]
Yay, that did the trick! Thank you so much for your efforts! 🍾

**Comment by lgiordani** [comment]
Good, I'm going to release it then. Thanks for the use case and for using punch.

**Comment by lgiordani** [comment]
Solved in version 3c9ae25d511ac07205dca5a471fcfb0ac015a2d0
